File: Sources/Ledgelings/SpriteAtlas.swift
import AppKit
import LedgelingsCore

/// A packed sprite sheet: the PNG and JSON that `spritetool pack` writes.
struct SpriteAtlas {
    struct Meta: Decodable {
        struct Rect: Decodable { let x, y, w, h: Int }
        struct Animation: Decodable { let frames: [String]; let fps: Double; let loop: Bool }
        let name: String
        let image: String
        let cell: [Int]
        let contentBox: [Int]
        let variants: [String]
        let palette: [String: String]?
        let frames: [String: Rect]
        let animations: [String: Animation]
    }

    /// One recolouring of the sheet, already cut into frames.
    struct Frames {
        fileprivate let meta: Meta
        fileprivate let images: [String: CGImage]

        /// The frame for an animation at `time` seconds in, with the given eye state.
        func frame(animation: String, time: Double, eyes: Eyes = .open) -> CGImage? {
            guard let anim = meta.animations[animation], !anim.frames.isEmpty else { return nil }
            let raw = Int(time * anim.fps)
            let index = anim.loop ? raw % anim.frames.count : min(raw, anim.frames.count - 1)
            let pose = anim.frames[index]
            return images["\(pose)_\(eyes.rawValue)"] ?? images[pose]
        }
    }

    enum LoadError: Error, CustomStringConvertible {
        case notFound(String), unreadable(String), badFrame(String)
        var description: String {
            switch self {
            case .notFound(let what): "sprite resource not found: \(what)"
            case .unreadable(let what): "cannot read sprite resource: \(what)"
            case .badFrame(let name): "atlas frame \(name) lies outside the sheet"
            }
        }
    }

    let meta: Meta
    private let sheet: CGImage

    var cellSize: CGSize { CGSize(width: meta.cell[0], height: meta.cell[1]) }
    /// Half the width of the square the creature is drawn inside, in sheet pixels.
    var bodyHalfSize: CGFloat { CGFloat(meta.contentBox[2]) / 2 }

    init(named name: String) throws {
        let directory = try Self.spritesDirectory()
        let jsonURL = directory.appendingPathComponent("\(name).json")
        guard let data = try? Data(contentsOf: jsonURL) else { throw LoadError.notFound(jsonURL.path) }
        meta = try JSONDecoder().decode(Meta.self, from: data)

        let imageURL = directory.appendingPathComponent(meta.image)
        guard let source = CGImageSourceCreateWithURL(imageURL as CFURL, nil),
              let image = CGImageSourceCreateImageAtIndex(source, 0, nil)
        else { throw LoadError.unreadable(imageURL.path) }
        sheet = image
        for (frame, r) in meta.frames where r.x + r.w > image.width || r.y + r.h > image.height {
            throw LoadError.badFrame(frame)
        }
    }

    /// The sheet cut into frames, recoloured around `body` when the atlas has a
    /// palette. Nil body means "as painted".
    func frames(body: RGB? = nil) -> Frames {
        let source = body.flatMap { recoloured(around: $0) } ?? sheet
        var cut: [String: CGImage] = [:]
        for (frame, r) in meta.frames {
            // CGImage cropping is top-left origin, the same as the atlas JSON.
            cut[frame] = source.cropping(to: CGRect(x: r.x, y: r.y, width: r.w, height: r.h))
        }
        return Frames(meta: meta, images: cut)
    }

    /// Swap the atlas palette for shades of `body`. Every other pixel -- the
    /// black eyes above all -- is left exactly as it was.
    private func recoloured(around body: RGB) -> CGImage? {
        guard let palette = meta.palette, !palette.isEmpty else { return nil }
        let targets: [String: RGB] = [
            "body": body, "light": body.mixed(with: .white, 0.36),
            "shade": body.mixed(with: .black, 0.17), "outline": body.mixed(with: .black, 0.76),
        ]
        let swaps = palette.compactMap { name, hex -> (RGB, RGB)? in
            guard let from = RGB(hex: hex), let to = targets[name] else { return nil }
            return (from, to)
        }
        let w = sheet.width, h = sheet.height
        guard let ctx = CGContext(data: nil, width: w, height: h, bitsPerComponent: 8, bytesPerRow: w * 4,
                                  space: CGColorSpace(name: CGColorSpace.sRGB)!,
                                  bitmapInfo: CGImageAlphaInfo.premultipliedLast.rawValue),
              let data = ctx.data else { return nil }
        ctx.draw(sheet, in: CGRect(x: 0, y: 0, width: w, height: h))
        let px = data.bindMemory(to: UInt8.self, capacity: w * h * 4)
        for i in stride(from: 0, to: w * h * 4, by: 4) where px[i + 3] == 255 {
            let here = RGB(r: px[i], g: px[i + 1], b: px[i + 2])
            if let (_, to) = swaps.first(where: { $0.0.isClose(to: here) }) {
                px[i] = to.r; px[i + 1] = to.g; px[i + 2] = to.b
            }
        }
        return ctx.makeImage()
    }

    /// Inside an .app the sheets sit in Contents/Resources/sprites; under
    /// `swift run` they are in the SwiftPM resource bundle. The main bundle is
    /// tried first ON PURPOSE: touching `Bundle.module` where that bundle does
    /// not exist (a hand-assembled .app) is a crash, not a nil.
    private static func spritesDirectory() throws -> URL {
        let exists = { (url: URL?) -> URL? in
            guard let dir = url?.appendingPathComponent("sprites", isDirectory: true),
                  FileManager.default.fileExists(atPath: dir.path) else { return nil }
            return dir
        }
        if let found = exists(Bundle.main.resourceURL) ?? exists(Bundle.module.resourceURL) { return found }
        throw LoadError.notFound("sprites/")
    }
}

/// An sRGB colour, 8 bits a channel, that round-trips through "#rrggbb".
struct RGB: Hashable {
    var r, g, b: UInt8

    static let white = RGB(r: 255, g: 255, b: 255)
    static let black = RGB(r: 0, g: 0, b: 0)

    init(r: UInt8, g: UInt8, b: UInt8) { self.r = r; self.g = g; self.b = b }

    init?(hex: String) {
        let text = hex.hasPrefix("#") ? String(hex.dropFirst()) : hex
        guard text.count == 6, let value = UInt32(text, radix: 16) else { return nil }
        self.init(r: UInt8(value >> 16 & 0xFF), g: UInt8(value >> 8 & 0xFF), b: UInt8(value & 0xFF))
    }

    var hex: String { String(format: "#%02x%02x%02x", r, g, b) }

    func mixed(with other: RGB, _ amount: Double) -> RGB {
        func mix(_ a: UInt8, _ b: UInt8) -> UInt8 { UInt8((Double(a) * (1 - amount) + Double(b) * amount).rounded()) }
        return RGB(r: mix(r, other.r), g: mix(g, other.g), b: mix(b, other.b))
    }

    /// Exact in practice; the slack absorbs a colour-managed decode being off by one.
    func isClose(to other: RGB) -> Bool {
        abs(Int(r) - Int(other.r)) <= 2 && abs(Int(g) - Int(other.g)) <= 2 && abs(Int(b) - Int(other.b)) <= 2
    }
}

File: win/Ledgelings/SpriteAtlas.cs
using System.Drawing;
using System.Globalization;
using System.Text.Json;
using System.Text.Json.Serialization;
using Ledgelings.Core;

namespace Ledgelings;

/// <summary>An sRGB colour, 8 bits a channel, that round-trips through "#rrggbb".</summary>
public readonly record struct RGB(byte R, byte G, byte B)
{
    public static readonly RGB White = new(255, 255, 255);
    public static readonly RGB Black = new(0, 0, 0);

    public static RGB? FromHex(string? hex)
    {
        if (hex is null) return null;
        var text = hex.StartsWith('#') ? hex[1..] : hex;
        if (text.Length != 6 || !uint.TryParse(text, NumberStyles.HexNumber, CultureInfo.InvariantCulture, out var value)) return null;
        return new RGB((byte)(value >> 16 & 0xFF), (byte)(value >> 8 & 0xFF), (byte)(value & 0xFF));
    }

    public string Hex => $"#{R:x2}{G:x2}{B:x2}";

    public RGB Mixed(RGB other, double amount)
    {
        byte Mix(byte a, byte b) => (byte)Math.Round(a * (1 - amount) + b * amount, MidpointRounding.AwayFromZero);
        return new RGB(Mix(R, other.R), Mix(G, other.G), Mix(B, other.B));
    }

    /// <summary>Exact in practice; the slack absorbs a colour-managed decode being off by one.</summary>
    public bool IsClose(RGB other) => Math.Abs(R - other.R) <= 2 && Math.Abs(G - other.G) <= 2 && Math.Abs(B - other.B) <= 2;

    public Color ToColor() => Color.FromArgb(R, G, B);
}

/// <summary>A packed sprite sheet: the PNG and JSON that <c>spritetool pack</c> writes.</summary>
public sealed class SpriteAtlas
{
    public sealed class Meta
    {
        public sealed record FrameRect(int X, int Y, int W, int H);
        public sealed record Animation(List<string> Frames, double Fps, bool Loop);
        public string Name { get; set; } = "";
        public string Image { get; set; } = "";
        public int[] Cell { get; set; } = { 32, 32 };
        public int[] ContentBox { get; set; } = { 5, 5, 22, 22 };
        public List<string> Variants { get; set; } = new();
        public Dictionary<string, string>? Palette { get; set; }
        public Dictionary<string, FrameRect> Frames { get; set; } = new();
        public Dictionary<string, Animation> Animations { get; set; } = new();
        /// <summary>What the creature is, and who its creatures are, when the sheet says.</summary>
        public string? Kind { get; set; }
        public List<Character>? Cast { get; set; }
        /// <summary>The body colour this species always wears, when the sheet chose one.</summary>
        public string? Colour { get; set; }
    }

    /// <summary>One recolouring of the sheet, already cut into frames.</summary>
    public sealed class Frames
    {
        private readonly Meta meta;
        private readonly Dictionary<string, Bitmap> images;
        internal Frames(Meta meta, Dictionary<string, Bitmap> images) { this.meta = meta; this.images = images; }

        /// <summary>The frame for an animation at <paramref name="time"/> seconds in, with the given eye state.</summary>
        public Bitmap? Frame(string animation, double time, Eyes eyes = Eyes.Open)
        {
            if (!meta.Animations.TryGetValue(animation, out var anim) || anim.Frames.Count == 0) return null;
            var raw = (int)(time * anim.Fps);
            var index = anim.Loop ? raw % anim.Frames.Count : Math.Min(raw, anim.Frames.Count - 1);
            var pose = anim.Frames[index];
            return images.TryGetValue($"{pose}_{eyes.Key()}", out var exact) ? exact : images.TryGetValue(pose, out var plain) ? plain : null;
        }
    }

    private static readonly JsonSerializerOptions jsonOptions = new()
    {
        PropertyNameCaseInsensitive = true, NumberHandling = JsonNumberHandling.AllowReadingFromString,
    };

    public Meta Info { get; }
    private readonly SpriteText.Image sheet;

    public Size CellSize => new(Info.Cell[0], Info.Cell[1]);
    /// <summary>Half the width of the square the creature is drawn inside, in sheet pixels.</summary>
    public double BodyHalfSize => Info.ContentBox[2] / 2.0;

    /// <summary>The folder of sheets shipped next to the exe.</summary>
    public static string SpritesDirectory => Path.Combine(AppContext.BaseDirectory, "sprites");

    /// <summary>A sheet shipped inside the app.</summary>
    public static SpriteAtlas Named(string name) => new(SpritesDirectory, name);

    /// <summary>A sheet in any folder: <c>&lt;name&gt;.json</c> next to the PNG it names.</summary>
    public SpriteAtlas(string directory, string name)
    {
        var jsonPath = Path.Combine(directory, name + ".json");
        if (!File.Exists(jsonPath)) throw new FileNotFoundException("sprite resource not found: " + jsonPath);
        Info = JsonSerializer.Deserialize<Meta>(File.ReadAllText(jsonPath), jsonOptions) ?? throw new InvalidDataException("empty atlas: " + jsonPath);
        var imagePath = Path.Combine(directory, Info.Image);
        if (!File.Exists(imagePath)) throw new FileNotFoundException("cannot read sprite resource: " + imagePath);
        sheet = Harden(PngIO.Read(imagePath));
        foreach (var (frame, r) in Info.Frames)
            if (r.X + r.W > sheet.Width || r.Y + r.H > sheet.Height) throw new InvalidDataException($"atlas frame {frame} lies outside the sheet");
    }

    /// <summary>Pixel art has 1-bit alpha: drop the faint pixels, make the rest solid.</summary>
    private static SpriteText.Image Harden(SpriteText.Image image)
    {
        var rgba = new byte[image.Rgba.Length];
        for (int i = 0; i < rgba.Length; i += 4)
        {
            if (image.Rgba[i + 3] < 128) continue;
            rgba[i] = image.Rgba[i]; rgba[i + 1] = image.Rgba[i + 1]; rgba[i + 2] = image.Rgba[i + 2]; rgba[i + 3] = 255;
        }
        return new SpriteText.Image(image.Width, image.Height, rgba);
    }

    /// <summary>The whole sheet as bytes, rows top to bottom, straight alpha.</summary>
    public SpriteText.Image Image() => sheet;

    /// <summary>The sheet cut into frames, recoloured around <paramref name="body"/> when the atlas has a
    /// palette. Null body means "as painted".</summary>
    public Frames MakeFrames(RGB? body = null)
    {
        var source = body is RGB colour ? Recoloured(colour) ?? sheet : sheet;
        var cut = new Dictionary<string, Bitmap>();
        foreach (var (frame, r) in Info.Frames)
            cut[frame] = PngIO.ToBitmap(source, new Rectangle(r.X, r.Y, r.W, r.H));
        return new Frames(Info, cut);
    }

    /// <summary>Swap the atlas palette for shades of <paramref name="body"/>. Every other pixel -- the
    /// black eyes above all -- is left exactly as it was.</summary>
    private SpriteText.Image? Recoloured(RGB body)
    {
        if (Info.Palette is null || Info.Palette.Count == 0) return null;
        var targets = new Dictionary<string, RGB>
        {
            ["body"] = body, ["light"] = body.Mixed(RGB.White, 0.36),
            ["shade"] = body.Mixed(RGB.Black, 0.17), ["outline"] = body.Mixed(RGB.Black, 0.76),
        };
        var swaps = new List<(RGB From, RGB To)>();
        foreach (var (name, hex) in Info.Palette)
            if (RGB.FromHex(hex) is RGB from && targets.TryGetValue(name, out var to)) swaps.Add((from, to));
        var rgba = (byte[])sheet.Rgba.Clone();
        for (int i = 0; i < rgba.Length; i += 4)
        {
            if (rgba[i + 3] != 255) continue;
            var here = new RGB(rgba[i], rgba[i + 1], rgba[i + 2]);
            foreach (var (from, to) in swaps)
                if (from.IsClose(here)) { rgba[i] = to.R; rgba[i + 1] = to.G; rgba[i + 2] = to.B; break; }
        }
        return new SpriteText.Image(sheet.Width, sheet.Height, rgba);
    }
}
